### DISCUSSION TITLE
NativeControlHost

### DISCUSSION THREAD

**Discussion by StefanKoell** [AvaloniaUI/Avalonia#9436] [Q&A]
Hi!
I'm trying to figure out how I can use a WinForms or ActiveX (specifically the RDP ActiveX) control in Avalonia UI using the NativeControlHost. I find several (closed) issues, stating that the NativeControlHost is available since 0.10.x. The link to the sample is broken and I can't find it. There's no docs or sample which shows how I can make use of this NativeControlHost.

Can anyone show me how this is done?

Thanks!

**Comment by jp2masa** [comment]
There's an example on ControlCatalog: https://github.com/AvaloniaUI/Avalonia/blob/058200fe641266ae920a42bfeb1f0cd7bc9fe985/samples/ControlCatalog/Pages/NativeEmbedPage.xaml.cs.

**Comment by StefanKoell** [comment]
Thanks for the link but I still don't know how to use it. Hard to decipher without any docs or comments. How/where would I create an instance of my ActiveX control and how do I add it to the "parent"/controls collection of the host?

**Reply by jp2masa** [reply]
On [`CreateNativeControlCore`](https://github.com/AvaloniaUI/Avalonia/blob/058200fe641266ae920a42bfeb1f0cd7bc9fe985/samples/ControlCatalog/Pages/NativeEmbedPage.xaml.cs#L67) you are given the parent handle. I don't have experience with ActiveX specifically, but you can, for example, create a Win32 window and set its parent as that handle or something similar.

**Comment by StefanKoell** [comment]
Still no luck :(
I created a minimal repro project:
[Test.zip](https://github.com/AvaloniaUI/Avalonia/files/10004685/Test.zip)

Maybe someone can have a look and figure out what I'm doing wrong.

Thanks!

**Comment by kekekeks** [comment]
It works **exactly** the same as HwndHost in WPF. In `CreateNativeControlCore` you are asked for an HWND handle of the window you want to embed.

Here is an example of embedding Win32 RichTextBox: https://github.com/AvaloniaUI/Avalonia/blob/058200fe641266ae920a42bfeb1f0cd7bc9fe985/samples/ControlCatalog.NetCore/NativeControls/Win/EmbedSample.Win.cs#L20-L32

**Reply by StefanKoell** [reply]
Thanks for the sample @kekekeks. I'm only (vaguely) familiar with the WindowsFormsHost in WPF. I'm not sure how to apply your code to my sample with the ActiveX control. Maybe I'm missing something. Can you see what I'm doing wrong?

**Reply by kekekeks** [reply]
To host _an ActiveX_ control one needs to implement `IOleControlSite`, `IOleClientSite`, `IOleInPlaceSite`, `ISimpleFrameSite`,           `IPropertyNotifySink`, `IReflect` and optionally `IVBGetControl` and `IGetVBAObject`.

We currently don't provide a built-in implementation.

**Reply by StefanKoell** [reply]
Too bad, I guess I then have to stick to WPF :(

Thanks for your help though. Much appreciated!

**Reply by StefanKoell** [reply]
I just tried wrapping the ActiveX in a WinForms control but this also doesn't work. Shouldn't that be possible?

**Reply by maxkatz6** [reply]
It should, but win32 and especially something outdated like ActiveX not the easiest things to integrate.
You can check:
- if same works in a plain winforms control
- if winforms controls works in avalonia without activex
- if everything works as expected if you add "app.manifest" file with supported os defined there
- if everything works as expected if you disable UseWindowsUIComposition ([this issue](https://github.com/AvaloniaUI/Avalonia/issues/8497#issue-1302978884) as an example shows where to set this option)

**Comment by StefanKoell** [comment]
I just wanted come back to this thread. I finally was able to play some more with Avalonia 11 preview and I'm happy to report that even using UseWindowsUIComposition I'm able to wrap an ActiveX control (such as RDP) into a WinForms control and successfully host it in AvaloniaUI. So far I haven't seen any obvious airspace or focus issues but I'm sure the devil is in the details ;)

Anyway, I'm glad I got this working and I'm looking forward to play more with AvaloniaUI. Thanks for all the help and the great work!

**Comment by cesarchefinho** [comment]
I need rhe same, but with IOS, because i Need to use an UITextField to receive an SMS code that is only possible in ios with native control.

Any Help?

**Reply by kekekeks** [reply]
https://github.com/AvaloniaUI/Avalonia/blob/master/samples/ControlCatalog.iOS/EmbedSample.iOS.cs
https://github.com/AvaloniaUI/Avalonia/blob/master/samples/ControlCatalog/Pages/NativeEmbedPage.xaml
https://github.com/AvaloniaUI/Avalonia/blob/master/samples/ControlCatalog/Pages/NativeEmbedPage.xaml.cs

**Comment by StayMeaty** [comment]
@StefanKoell I'm also wrapping up an AxMstscRdp element in a NativeControlHost and it works fine but I actually am affected by the airspace problem in Avalonia 11.0.4, regardless of UseWindowsUIComposition. Can you by any chance provide a repo that shows how you do it?

**Reply by StefanKoell** [reply]
I'm not doing anything special. I'm wondering where you see the airspace issue. I also have to admit, that I haven't really battle tested my implementation. I just did a small PoC implementation. Can you share some code/sample project?

**Reply by StayMeaty** [reply]
I put the ActiveX element into a WinForms form and this form into a NativeControlHost control. Everything else I add - no matter where in the windows visual tree - is somehow always behind the AX element and so not visible. I'm still trying a few things, would otherwise finish a demo repo

**Reply by StefanKoell** [reply]
I see. I'm not putting anything "on top" of the control.

I've seen this post:
https://github.com/AvaloniaUI/Avalonia/discussions/10683#discussioncomment-5337864

Not sure if this can be used in your scenario but it shows how you can put other elements on top of a NativeControlHost.
